New rule: Color contrast enhanced (#1526)
act-rules/act-rules.github.io: _rules/text-contrast-enhanced-09o5cg.md

  wcag20:1.4.6: # Contrast (Enhanced)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:G17: # Ensuring that a contrast ratio of at least 7:1 exists between text (and images of text) and background behind the text
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:G18: # Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - Accessibility Tree
  - DOM Tree
  - CSS Styling
  - Language
acknowledgments:
  authors:
    - Jean-Yves Moyen
---

## Applicability

This rule applies to any [visible][] character in a [text node][] that is a [child][] in the [flat tree][] of an HTML element, except if the [text node][] has an [ancestor][] in the [flat tree][] for which one of the following is true:

- **widget**: the ancestor has a [semantic role][] that inherits from `widget`; or

> - **disabled label**: the ancestor is used in the [accessible name][] of a `widget` that is [disabled][]; or

Reviewer: ```suggestion
- **disabled label**: the ancestor is used in the [accessible name][] of a **widget** that is [disabled][]; or
```

I believe the "widget" in this sentence is referring to the widget condition that was expressed in the previous sentence. If it is referring to the "widget" semantic role then it needs to be rephrased because semantic roles do not have accessible names (elements do).

Author: 🤔 Actually wondering if it's worth creating a definition of "semantic widget", the same way we have "semantic link" (in several PRs now).
WDYT?

Reviewer: I think it is worth it 👍

## Accessibility Support

- Different browsers have different levels of support for CSS. This can cause contrast issues in one browser that do not appear in another. Because of that, this rule can produce different results depending on the browser that is used. For example, a text that is positioned using CSS transform may be on a different background in a browser that does not support CSS transform.
- Implementation of [Presentational Roles Conflict Resolution][] varies from one browser or assistive technology to another. Depending on this, some elements can have a [semantic role][] of `none` and fail this rule with some technology but users of other technologies would not experience any accessibility issue.

## Background

Passing this rule does not mean that the text has sufficient color contrast. If all background pixels have a low contrast with all foreground pixels, the success criterion is guaranteed to not be satisfied. When some pixels have sufficient contrast, and others do not, legibility should be considered. There is no clear method for determining legibility, which is why this is out of scope for this rule.

> When the text color or background color is not specified in the web page, colors from other [origins][] will be used. Testers must ensure colors are not affected by styles from a [user origin][], such as a custom style sheet. Contrast issues cause by specifying the text color but not the background or vise versa, must be tested separately from this rule.

Reviewer: ```suggestion
When the text color or background color is not specified in the web page, colors from other [origins][] will be used. Testers must ensure colors are not affected by styles from a [user origin][], such as a custom style sheet. Contrast issues caused by specifying the text color but not the background or vice versa, must be tested separately from this rule.
```

- [CSS Scoping Module Level 1 (Editor's Draft)](https://drafts.csswg.org/css-scoping/)

## Test Cases

### Passed

#### Passed Example 1

This dark gray text has a contrast ratio of 12.6:1 on the white background.

```html
<p style="color: #333; background: #FFF;">
	Some text in a human language
</p>
```

#### Passed Example 2

This dark gray text has a contrast ratio between 12.6:1 and 7:1 on the white to light blue gradient background.

```html
<p style="color: #333; background: linear-gradient(to right, #FFF, #BBF); width: 500px;">
	Some text in a human language
</p>
```

#### Passed Example 3

This light gray text has a contrast ratio between 18:1 and 7:1 on the background image.

```html
<p
	style="color: #EEE; height:50px; padding-top:15px; background: #000 no-repeat -20px -20px url('/test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Passed Example 4

This 18pt large black text has a contrast ratio of 4.6:1 on the gray background.

```html
<p style="color: #000; font-size:18pt; background: #777;">
	Some text in a human language
</p>
```

#### Passed Example 5

> This 14pt bold black text has a contrast ratio of 3.6:1 on the gray background.

Reviewer: ```suggestion
This 14pt bold black text has a contrast ratio of 4.6:1 on the gray background.
```

	Some text in English
</p>
```

#### Failed Example 7

This black text with 60% opacity has a contrast ratio of 5.7:1 on the white background.

```html
<div style="background: #FFF">
	<p style="color: #000; opacity: .6">
		Some text in English
	</p>
</div>
```

#### Failed Example 8

This gray text has a contrast ratio of 5.7:1 on the white background in a shadow DOM tree.

```html
<p style="color: #666; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

#### Failed Example 9

This semi-transparent gray text has a contrast ratio between 2.6:1 and 5.4:1 on the black and white background. The light gray text is compared to the white section of the background and the dark gray text is compared to the black section of the background.

```html
<style>
	#backgroundSplit {
		color: rgba(90, 90, 90, 0.9);
		background-position: top 0 left 0;
		background-image: linear-gradient(90deg, transparent, transparent 3.3em, black 3.3em, black 6em);
		padding: 0 1em;
	}
</style>
<span id="backgroundSplit">
	Hello world
</span>
```

#### Failed Example 10

> The first `p` element has a contrast ratio of 12.6:1. The second `p` element contains an example of the Helvetica font, which has a contrast ratio of 6.4:1. Because this provides information, and not only for aesthetic purposes, this is not considered [purely decorative][].

Reviewer: ```suggestion
The first `p` element has a contrast ratio of 12.6:1. The second `p` element, which contains an example of the Helvetica font, has a contrast ratio of 6.4:1. Because this provides information, and not only for aesthetic purposes, this is not considered [purely decorative][].
```
